src/dmd/typinf.d

```diff
@@ -72,40 +72,43 @@ extern (C++) void genTypeInfo(const ref Loc loc, Type torig, Scope* sc)
         // ClassInfos are generated as part of ClassDeclaration codegen
         const isUnqualifiedClassInfo = (t.ty == Tclass && !t.mod);
 
-        // generate a COMDAT for other TypeInfos not available as builtins in
-        // druntime
+        // Only generate TypeInfos not available as builtins in druntime.
         if (!isUnqualifiedClassInfo && !builtinTypeInfo(t))
-        {
-            if (sc) // if in semantic() pass
-            {
-                // Find module that will go all the way to an object file
-                Module m = sc._module.importedFrom;
-                m.members.push(t.vtinfo);
-            }
-            else // if in obj generation pass
-            {
-                toObjFile(t.vtinfo, global.params.multiobj);
-            }
-        }
+            tinfo = t.vtinfo;
     }
     if (!torig.vtinfo)
         torig.vtinfo = t.vtinfo; // Types aren't merged, but we can share the vtinfo's
     assert(torig.vtinfo);
+
+    return tinfo;
 }
 
 /****************************************************
- * Gets the type of the `TypeInfo` object associated with `t`
+ * Gets the type of the `TypeInfo` object associated with `t`.
+ * If called from the semantic pass, the generated `TypeInfo` is
+ * also appended to the root module's members array so that it
+ * will end up being written to object file.
  * Params:
  *      loc = the location for reporting line nunbers in errors
  *      t   = the type to get the type of the `TypeInfo` object for
  *      sc  = the scope
  * Returns:
  *      The type of the `TypeInfo` object associated with `t`
  */
-extern (C++) Type getTypeInfoType(const ref Loc loc, Type t, Scope* sc)
+Type getTypeInfoType(const ref Loc loc, Type t, Scope* sc)
 {
     assert(t.ty != Terror);
-    genTypeInfo(loc, t, sc);
+    if (!(sc.flags & SCOPE.ctfe) && !global.params.useTypeInfo)
+    {
+        .error(loc, "`TypeInfo` cannot be used with -betterC");
+        fatal();
+    }
+    if (auto tinfo = genTypeInfo(loc, t))
+    {
+        // Find module that will go all the way to an object file
+        Module m = sc._module.importedFrom;
+        m.members.push(t.vtinfo);
```

Comment:
FWIW, I don't assign any TypeInfos here to modules anymore for LDC. Instead, I emit struct and ad-hoc TypeInfos into each referencing object file, solely from the glue layer. Besides potentially less inter-object-file dependencies and increased optimization potential, this also takes care of all CTFE / speculative instantiation problems.